// @nrwl/angular/src/generators/application/lib/add-unit-test-runner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addUnitTestRunner = void 0;
const tslib_1 = require("tslib");
const jest_1 = require("@nrwl/jest");
const test_runners_1 = require("../../../utils/test-runners");
const karma_project_1 = require("../../karma-project/karma-project");
function addUnitTestRunner(host, options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        if (options.unitTestRunner === test_runners_1.UnitTestRunner.Jest) {
            yield (0, jest_1.jestProjectGenerator)(host, {
                project: options.name,
                setupFile: 'angular',
                supportTsx: false,
                skipSerializers: false,
                skipPackageJson: options.skipPackageJson,
            });
        }
        else if (options.unitTestRunner === test_runners_1.UnitTestRunner.Karma) {
            yield (0, karma_project_1.default)(host, {
                project: options.name,
                skipFormat: true,
            });
        }
    });
}
exports.addUnitTestRunner = addUnitTestRunner;
