// @nrwl/angular/src/utils/versions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.typesNodeVersion = exports.typesJasminewd2Version = exports.typesJasmineVersion = exports.jasmineSpecReporterVersion = exports.jasmineCoreVersion = exports.karmaJasmineHtmlReporterVersion = exports.karmaJasmineVersion = exports.karmaCoverageVersion = exports.karmaChromeLauncherVersion = exports.karmaVersion = exports.protractorVersion = exports.jestPresetAngularVersion = exports.tsNodeVersion = exports.autoprefixerVersion = exports.postcssUrlVersion = exports.postcssPresetEnvVersion = exports.postcssImportVersion = exports.postcssVersion = exports.tailwindVersion = exports.angularEslintVersion = exports.tsLibVersion = exports.angularJsVersion = exports.zoneJsVersion = exports.rxjsVersion = exports.ngrxVersion = exports.ngPackagrVersion = exports.angularDevkitVersion = exports.angularVersion = exports.nxVersion = void 0;
exports.nxVersion = require('../../package.json').version;
exports.angularVersion = '~14.2.0';
exports.angularDevkitVersion = '~14.2.0';
exports.ngPackagrVersion = '~14.2.0';
exports.ngrxVersion = '~14.0.0';
exports.rxjsVersion = '~7.5.0';
exports.zoneJsVersion = '~0.11.4';
exports.angularJsVersion = '1.7.9';
exports.tsLibVersion = '^2.3.0';
exports.angularEslintVersion = '~14.0.0';
exports.tailwindVersion = '^3.0.2';
exports.postcssVersion = '^8.4.5';
exports.postcssImportVersion = '~14.1.0';
exports.postcssPresetEnvVersion = '~7.5.0';
exports.postcssUrlVersion = '~10.1.3';
exports.autoprefixerVersion = '^10.4.0';
exports.tsNodeVersion = '10.9.1';
exports.jestPresetAngularVersion = '~12.2.0';
exports.protractorVersion = '~7.0.0';
exports.karmaVersion = '~6.4.0';
exports.karmaChromeLauncherVersion = '~3.1.0';
exports.karmaCoverageVersion = '~2.2.0';
exports.karmaJasmineVersion = '~5.1.0';
exports.karmaJasmineHtmlReporterVersion = '~2.0.0';
exports.jasmineCoreVersion = '~4.2.0';
exports.jasmineSpecReporterVersion = '~7.0.0';
exports.typesJasmineVersion = '~4.0.0';
exports.typesJasminewd2Version = '~2.0.3';
exports.typesNodeVersion = '16.11.7';
