// @nrwl/angular/testing/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.readFirst = exports.readAll = exports.time = exports.getTestScheduler = exports.hot = exports.cold = void 0;
const jasmine_marbles_1 = require("jasmine-marbles");
/**
 * @deprecated Import from 'jasmine-marbles' instead. Will be removed in Nx v15.
 */
exports.cold = jasmine_marbles_1.cold;
/**
 * @deprecated Import from 'jasmine-marbles' instead. Will be removed in Nx v15.
 */
exports.hot = jasmine_marbles_1.hot;
/**
 * @deprecated Import from 'jasmine-marbles' instead. Will be removed in Nx v15.
 */
exports.getTestScheduler = jasmine_marbles_1.getTestScheduler;
/**
 * @deprecated Import from 'jasmine-marbles' instead. Will be removed in Nx v15.
 */
exports.time = jasmine_marbles_1.time;
var testing_utils_1 = require("./src/testing-utils");
Object.defineProperty(exports, "readAll", { enumerable: true, get: function () { return testing_utils_1.readAll; } });
Object.defineProperty(exports, "readFirst", { enumerable: true, get: function () { return testing_utils_1.readFirst; } });
