// @nrwl/cypress/src/generators/cypress-component-project/cypress-component-project.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addToCacheableOperations = exports.cypressComponentProject = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nrwl/devkit");
const cypress_version_1 = require("../../utils/cypress-version");
const versions_1 = require("../../utils/versions");
function cypressComponentProject(tree, options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const cyVersion = (0, cypress_version_1.installedCypressVersion)();
        if (cyVersion && cyVersion < 10) {
            throw new Error('Cypress version of 10 or higher is required to use component testing. See the migration guide to upgrade. https://nx.dev/cypress/v10-migration-guide');
        }
        const projectConfig = (0, devkit_1.readProjectConfiguration)(tree, options.project);
        const installDepsTask = updateDeps(tree);
        addProjectFiles(tree, projectConfig, options);
        addTargetToProject(tree, projectConfig, options);
        addToCacheableOperations(tree);
        if (!options.skipFormat) {
            yield (0, devkit_1.formatFiles)(tree);
        }
        return () => {
            installDepsTask();
        };
    });
}
exports.cypressComponentProject = cypressComponentProject;
function updateDeps(tree) {
    const devDeps = {
        '@cypress/webpack-dev-server': versions_1.cypressWebpackVersion,
        'html-webpack-plugin': versions_1.webpackHttpPluginVersion,
        cypress: versions_1.cypressVersion,
    };
    return (0, devkit_1.addDependenciesToPackageJson)(tree, {}, devDeps);
}
function addProjectFiles(tree, projectConfig, options) {
    (0, devkit_1.generateFiles)(tree, (0, devkit_1.joinPathFragments)(__dirname, 'files'), projectConfig.root, Object.assign(Object.assign({}, options), { projectRoot: projectConfig.root, offsetFromRoot: (0, devkit_1.offsetFromRoot)(projectConfig.root), ext: '' }));
}
function addTargetToProject(tree, projectConfig, options) {
    projectConfig.targets['component-test'] = {
        executor: '@nrwl/cypress:cypress',
        options: {
            cypressConfig: (0, devkit_1.joinPathFragments)(projectConfig.root, 'cypress.config.ts'),
            testingType: 'component',
        },
    };
    (0, devkit_1.updateProjectConfiguration)(tree, options.project, projectConfig);
}
function addToCacheableOperations(tree) {
    (0, devkit_1.updateJson)(tree, 'nx.json', (json) => {
        var _a, _b, _c, _d, _e, _f, _g;
        return (Object.assign(Object.assign({}, json), { tasksRunnerOptions: Object.assign(Object.assign({}, json.tasksRunnerOptions), { default: Object.assign(Object.assign({}, (_a = json.tasksRunnerOptions) === null || _a === void 0 ? void 0 : _a.default), { options: Object.assign(Object.assign({}, (_c = (_b = json.tasksRunnerOptions) === null || _b === void 0 ? void 0 : _b.default) === null || _c === void 0 ? void 0 : _c.options), { cacheableOperations: Array.from(new Set([
                            ...((_g = (_f = (_e = (_d = json.tasksRunnerOptions) === null || _d === void 0 ? void 0 : _d.default) === null || _e === void 0 ? void 0 : _e.options) === null || _f === void 0 ? void 0 : _f.cacheableOperations) !== null && _g !== void 0 ? _g : []),
                            'component-test',
                        ])) }) }) }) }));
    });
}
exports.addToCacheableOperations = addToCacheableOperations;
